src/app/libraries/token_generator_lib.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+class token_generator_lib extends Library {
+
+    public function __construct() {
+        if (session_status() == PHP_SESSION_NONE) {
+            session_start();
+        }
+    }
+    /**
+     * Generates a new CSRF token
+     * @return string Base64 encoded token
+     */
+    public static function getNewCSRFToken() {
+        return base64_encode(openssl_random_pseudo_bytes(12));
+    }
+
+    /**
+     * Returns the currently set CSRF token
+     * @return string|bool The token
+     */
+    public function getCurrentCSRFToken() {
+        if (isset($_SESSION['csrf_token'])) {
+            return $_SESSION['csrf_token'];
+        }
+        return false;
+    }
+
+    /**
+     * Generates a new CSRF token and sets it as the current token
+     * @return string The token
+     */
+    public function setNewCSRFToken() {
+        $csrfToken = self::getNewCSRFToken();
+        $_SESSION['csrf_token'] = $csrfToken;
+        return $csrfToken;
+    }
+
+    /**
+     * Checks whether the current CSRF token from the POST data is correct. If not, returns a
+     * 400 HTTP response code, loads the 400 view and quits.
+     */
+    public function checkCSRFToken($value) {
+        if ($value !== $this->getCurrentCSRFToken()) {
+            $this->load_library('http_lib', 'http');
+            $this->http->response_code(400, true);
```

Comment:
The default for second parameter is true, I think. We need to decide on some convention for this.